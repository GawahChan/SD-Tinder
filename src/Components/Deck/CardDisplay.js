import React, { useState } from "react";
import PropTypes from "prop-types";
import {
  CardContents,
  CardOverlay,
  CardContent,
  CardButtonContainer,
  CardSubtitleContainer,
  Container
} from "./style";

import {
  CardInfoButton,
  CardLikedButton,
  CardDislikedButton,
  CardSuperlikedButton
} from "../../Common/Styled/Icons.styles";
import { Header, Subtitle } from "../../Common/Styled/Typography.styles";

const CardDisplay = ({ stuntDouble, animateAction, action }) => {
  const [actionType, setActionType] = useState("");
  const [displayBio, setDisplayBio] = useState(false);

  const receiveAction = (id, actionType) => {
    animateAction(id, actionType);
    setActionType(actionType);
  };

  const toggleBio = () => setDisplayBio(!displayBio);

  const overlay = actionType ? actionType.toUpperCase() : action.toUpperCase();

  return (
    <CardContents actionColour={overlay} img={stuntDouble.img}>
      <CardOverlay actionColour={overlay}>{overlay}</CardOverlay>
      <CardContent>
        <Container>
          <Header>
            {stuntDouble.name}, {stuntDouble.age}
          </Header>
          <CardSubtitleContainer>
            <Subtitle>{stuntDouble.location}</Subtitle>
            <CardInfoButton onTap={toggleBio} />
          </CardSubtitleContainer>
          {displayBio && <Subtitle>{stuntDouble.bio}</Subtitle>}
        </Container>
        <Container>
          <CardButtonContainer>
            <CardDislikedButton
              onClick={() => receiveAction(stuntDouble.id, "disliked")}
            />
            <CardSuperlikedButton
              onClick={() => receiveAction(stuntDouble.id, "superliked")}
            />
            <CardLikedButton
              onClick={() => receiveAction(stuntDouble.id, "liked")}
            />
          </CardButtonContainer>
        </Container>
      </CardContent>
    </CardContents>
  );
};

CardDisplay.propTypes = {
  stuntDouble: PropTypes.object,
  animateAction: PropTypes.func,
  action: PropTypes.string
};
export default CardDisplay;
