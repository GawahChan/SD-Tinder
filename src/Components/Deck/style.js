import styled from "styled-components";
import { motion } from "framer-motion";

export const CardsContainer = styled(motion.div)`
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  min-height: 100vh;
  flex: 1;
`;
// eslint-disable-next-line
export const CardContainer = styled(motion.div).attrs(props => ({
  style: {
    boxShadow:
      props.actionColour === "liked"
        ? "0px 0px 20px #15ff00"
        : props.actionColour === "disliked"
        ? "0px 0px 20px #ff0000"
        : props.actionColour === "superliked"
        ? "0 0 20px #02fdfd"
        : "0px 0px 10px #080808",

    border:
      props.actionColour === "liked"
        ? "1px solid #15ff00"
        : props.actionColour === "disliked"
        ? "1px solid #ff0000"
        : props.actionColour === "superliked"
        ? "1px solid #02fdfd"
        : "none"
  }
}))`
  position: absolute;
  border-radius: 10px;
  overflow: hidden;
  cursor: pointer;

  width: 320px;
  height: 480px;

  transition: box-shadow 0.5s, border 0.5s;
  will-change: transform;
`;

export const CardContents = styled(motion.div)`
  width: 100%;
  height: 100%;

  display: flex;
  flex-direction: column;
  justify-content: flex-end;

  background-image: ${props =>
    `url(${require(`../../Common/Profile/${props.img}.jpg`)})`};
  background-position: top;
  background-size: cover;

  transition: box-shadow 0.5s;
  will-change: transform;
`;

export const CardContent = styled.div`
  padding: 10px;
  box-sizing: border-box;
  color: #dddddd;
  background: linear-gradient(
    rgba(0, 0, 0, 0),
    rgba(0, 0, 0, 0.7),
    rgba(0, 0, 0, 0.8),
    rgba(0, 0, 0, 1)
  );
`;

// eslint-disable-next-line
export const CardOverlay = styled.h1.attrs(props => ({
  style: {
    color:
      props.actionColour === "LIKED"
        ? "#15ff00"
        : props.actionColour === "DISLIKED"
        ? "#ff0000"
        : props.actionColour === "SUPERLIKED"
        ? "#02fdfd"
        : "black"
  }
}))`
  height: 30%;
  width: 100%;

  display: flex;
  justify-content: center;
  align-items: center;

  font-size: 3rem;
  font-weight: 500;
`;

export const CardSubtitleContainer = styled.div`
  display: flex;
  align-items: center;
  justify-content: space-between;
  box-sizing: border-box;
  padding-top: 5px;
  padding-bottom: 5px;
`;

export const CardButtonContainer = styled.div`
  display: flex;
  align-items: center;
  justify-content: center;
`;

export const Container = styled.div`
  box-sizing: border-box;
  padding: 5px;
`;
