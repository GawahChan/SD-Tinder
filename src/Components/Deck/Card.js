import React, { useState } from "react";
import PropTypes from "prop-types";
import { CardContainer } from "./style";
import CardDisplay from "./CardDisplay";

import {
  INITIAL_CARD_POSITION,
  REMOVE_LIKED_ANIMATION,
  REMOVE_DISLIKED_ANIMATION,
  REMOVE_SUPERLIKED_ANIMATION,
  DISPLAY_SUPERLIKED_BOUNDARY,
  DISPLAY_LIKED_BOUNDARY,
  DISPLAY_DISLIKED_BOUNDARY,
  REMOVE_SUPERLIKED_BOUNDARY,
  REMOVE_LIKED_BOUNDARY,
  REMOVE_DISLIKED_BOUNDARY
} from "../../Constants";

const Card = ({ stuntDouble, handleAction }) => {
  const [action, setAction] = useState("");
  const [removeCard, setRemoveCard] = useState(INITIAL_CARD_POSITION);

  const animateAction = (id, action) => {
    let removeCard =
      action === "liked"
        ? REMOVE_LIKED_ANIMATION
        : action === "disliked"
        ? REMOVE_DISLIKED_ANIMATION
        : action === "superliked"
        ? REMOVE_SUPERLIKED_ANIMATION
        : INITIAL_CARD_POSITION;

    setAction(action);
    setRemoveCard(removeCard);

    handleAction(id, action);
  };

  const handleDrag = info => {
    let xPosition = info.point.x;
    let yPosition = info.point.y;

    let action =
      xPosition > DISPLAY_LIKED_BOUNDARY
        ? "liked"
        : xPosition < DISPLAY_DISLIKED_BOUNDARY
        ? "disliked"
        : yPosition < DISPLAY_SUPERLIKED_BOUNDARY
        ? "superliked"
        : "";

    setAction(action);
  };

  const handleDragEnd = (id, info) => {
    let xPosition = info.point.x;
    let yPosition = info.point.y;

    return xPosition > REMOVE_LIKED_BOUNDARY
      ? animateAction(id, "liked")
      : xPosition < REMOVE_DISLIKED_BOUNDARY
      ? animateAction(id, "disliked")
      : yPosition < REMOVE_SUPERLIKED_BOUNDARY
      ? animateAction(id, "superliked")
      : setAction("");
  };

  return (
    <CardContainer
      drag
      dragElastic={0.8}
      dragConstraints={{ top: 0, right: 0, bottom: 0, left: 0 }}
      onDrag={(e, info) => handleDrag(info)}
      onDragEnd={(e, info) => handleDragEnd(stuntDouble.id, info)}
      initial={{ opacity: 0 }}
      animate={{ opacity: 1 }}
      exit={removeCard}
      actionColour={action}
    >
      <CardDisplay
        stuntDouble={stuntDouble}
        animateAction={animateAction}
        action={action}
      />
    </CardContainer>
  );
};

Card.propTypes = {
  stuntDouble: PropTypes.object,
  handleAction: PropTypes.func
};

export default Card;
