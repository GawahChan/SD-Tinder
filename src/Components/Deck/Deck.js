import React, { Component } from "react";
import { Redirect } from "react-router-dom";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import { bindActionCreators } from "redux";
import {
  removeStuntDouble,
  addStuntDouble
} from "../../Store/Actions/actionTypes";

import Card from "./Card";
import { HOME_URL } from "../../Constants";

import { AnimatePresence } from "framer-motion";
import { CardsContainer } from "./style";

class Deck extends Component {
  handleAction = (id, action) => {
    this.props.addStuntDouble(id, action);
    this.props.removeStuntDouble(id);
  };

  render() {
    const { stuntDoublesList } = this.props;
    return (
      <div>
        {stuntDoublesList.length > 0 ? (
          <CardsContainer>
            <AnimatePresence>
              {stuntDoublesList.map(stuntDouble => {
                return (
                  <Card
                    key={stuntDouble.id}
                    stuntDouble={stuntDouble}
                    handleAction={this.handleAction}
                  />
                );
              })}
            </AnimatePresence>
          </CardsContainer>
        ) : (
          <Redirect to={HOME_URL} />
        )}
      </div>
    );
  }
}

Deck.propTypes = {
  stuntDoublesList: PropTypes.array,
  removeStuntDouble: PropTypes.func,
  addStuntDOuble: PropTypes.func
};

const mapStateToProps = state => {
  return {
    stuntDoublesList: state.stuntDoubles
  };
};

const mapDispatchToProps = dispatch => {
  return bindActionCreators({ removeStuntDouble, addStuntDouble }, dispatch);
};

export default connect(mapStateToProps, mapDispatchToProps)(Deck);
