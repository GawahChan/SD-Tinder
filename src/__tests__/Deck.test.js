import React from 'react';
import { shallow } from 'enzyme';
import Deck from '../Components/Deck/Deck';

describe('Deck component', () => {
    const deck = shallow(<Deck />);

    it('renders correctly', () => {
        expect(deck).toMatchSnapshot();
    })

    it('initializes the liked `state` with an empty array', () => {
        expect(deck.state('liked')).toEqual([]);
    })

    describe('user Likes a profile', () => {
        const id = 0;
        const liked = { id: 0, name: 'Mila Kunis', age: '36', location: 'LA, USA', bio: 'Actress' }
        beforeEach(() => {
            deck.instance().liked(id);
        });

        it('adds the liked profile to the state `liked`', () => {
            expect(deck.state('liked')).toHaveLength(1);
            expect(deck.state('liked')).toContainEqual(liked);
        });

        it('removes the liked profile from the state `stuntDoubleList`', () => {
            expect(deck.state('stuntDoubleList')).not.toContain(liked)
        });
    })
})