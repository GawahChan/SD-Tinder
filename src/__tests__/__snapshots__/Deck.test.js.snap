// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Deck component renders correctly 1`] = `
<div
  className="app"
>
  <h1>
    Stunt Double App
  </h1>
  <Card
    age="36"
    bio="Actress"
    id={0}
    key="0"
    liked={[Function]}
    location="LA, USA"
    name="Mila Kunis"
  />
  <Card
    age="29"
    bio="Actress"
    id={1}
    key="1"
    liked={[Function]}
    location="Kentucky, USA"
    name="Jennifer Lawrence"
  />
  <Card
    age="29"
    bio="Actress"
    id={2}
    key="2"
    liked={[Function]}
    location="Oxford, UK"
    name="Emma Watson"
  />
</div>
`;
